src/core/load-fragment.c

```diff
@@ -5633,62 +5633,76 @@ int config_parse_cgroup_socket_bind(
                 void *data,
                 void *userdata) {
         _cleanup_free_ CGroupSocketBindItem *item = NULL;
-        const char *user_port;
+        _cleanup_strv_free_ char **parts = NULL;
+        int af = AF_UNSPEC, ip_protocol = 0, r;
         uint16_t nr_ports = 0, port_min = 0;
         CGroupSocketBindItem **head = data;
-        _cleanup_free_ char *word = NULL;
-        int af, r;
+        size_t parts_len;
 
         if (isempty(rvalue)) {
                 cgroup_context_remove_socket_bind(head);
                 return 0;
         }
 
-        r = extract_first_word(&rvalue, &word, ":", 0);
-        if (r == -ENOMEM)
+        parts = strv_split(rvalue, ":");
+        if (!parts)
                 return log_oom();
-        if (r <= 0) {
-                log_syntax(unit, LOG_WARNING, filename, line, r,
-                           "Unable to parse %s= assignment, ignoring: %s", lvalue, rvalue);
+
+        parts_len = strv_length(parts);
+        if (parts_len == 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Unable to parse %s= assignment, ignoring: %s.", lvalue, rvalue);
                 return 0;
         }
 
-        if (rvalue) {
-                af = af_from_ipv4_ipv6(word);
+        switch (parts_len) {
+        case 3:
+                /* IP protocol. */
+                ip_protocol = ip_protocol_from_tcp_udp(parts[1]);
+                if (ip_protocol < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, 0,
+                                   "Invalid IP protocol \"%s\", ignoring.", parts[1]);
+                        return 0;
+                }
+                _fallthrough_;
+        case 2:
+                /* Address family. */
+                af = af_from_ipv4_ipv6(parts[0]);
```

Comment:
I think we should allow the second and third parts to be empty: i.e. insert `if (!isempty(parts[0]))` here, and similarly above and below.

Also ties in nicely to the idea of using a shared helper… We want the same rules in both places.

Reply:
@keszybz

Introduced a shared helper, `parse_socket_bind_item`: it's more flexible than the previous version and addresses your feedback (all string tokens are optional). Please see corresponding UTs with valid and invalid examples.

Also manually tested with disk image, e.g.
```
[root@image ~]# SYSTEMD_SECCOMP=0 systemd-nspawn -D/ --volatile=yes --property=SocketBindAllow=udp:1500 --quiet -- bpftool m dump name sd_bind_allow
[{
        "key": 0,
        "value": {
            "address_family": 0,
            "protocol": 17,
            "nr_ports": 1,
            "port_min": 1500
        }
    }
]
[root@image ~]# 
```